pyembed/src/importer.rs

```diff
@@ -900,7 +904,219 @@ impl OxidizedFinder {
+            _ => false,
+        }
+    }
+
+    fn find_spec_impl(
+        &self,
+        py: Python,
+        fullname: &str,
+        target: Option<PyModule>,
+    ) -> PyResult<Option<PyObject>> {
+        if !_PathEntryFinder::is_visible(self.package(py), fullname) {
+            return Ok(Some(py.None()));
+        }
+        self.finder(py)
+            .call_method(py, "find_spec", (fullname, self.path(py), target), None)
+            .map(|spec| if spec == py.None() { None } else { Some(spec) })
+    }
+
+    fn iter_modules_impl(&self, py: Python, prefix: &str) -> PyResult<PyList> {
+        let state = self.finder(py).cast_as::<OxidizedFinder>(py)?.state(py);
+        let modules = state.get_resources_state().pkgutil_modules_infos(
+            py,
+            Some(prefix.to_string()),
+            state.optimize_level,
+            |resource| _PathEntryFinder::is_visible(self.package(py), &resource.name),
+        );
+        // unwrap() is safe because pkgutil_modules_infos returns a PyList cast
+        // into a PyObject.
+        Ok(modules?.cast_into(py).unwrap())
+    }
+
+    // Transform b"pkg/mod" into "pkg.mod" on behalf of `OxidizedFinder::path_hook`.
+    // Raise an `ImportError` if decoding `path` fails.
+    fn parse_path_to_pkg(py: Python, path: &std::path::Path) -> PyResult<String> {
+        let mut pkg = Vec::new();
+        for part in path.iter() {
+            pkg.push(
+                // Use Python's filesystem encoding to ensure correctly
+                // round-tripping the str Python package name from the
+                // potentially non-Unicode path.
+                // https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_170
+                // https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file
+                crate::conversion::path_to_pyobject(py, std::path::Path::new(part))?
+                    .cast_as::<PyString>(py)?
+                    .to_string(py)
+                    .map_err(|mut unicode_err| {
+                        // Need to raise an ImportError if can't decode. For
+                        // clarity, we also attach the UnicodeDecodeError.
+                        // https://docs.python.org/3/reference/import.html#path-entry-finders
+                        let mut imp_err = PyErr::new::<ImportError, _>(py, "cannot decode path");
+                        let imp_exc = imp_err.instance(py);
+                        if let Err(err) = imp_exc.setattr(py, "__suppress_context__", true) {
+                            err
+                        } else if let Err(err) =
+                            imp_exc.setattr(py, "__cause__", unicode_err.instance(py))
+                        {
+                            err
+                        } else {
+                            imp_err
+                        }
+                    })?
+                    .into_owned(),
+            );
+        }
```

Comment:
I'll have to think about this a bit.

We know that `/` and `\` are the only valid directory separators. So assuming we get a relative path, we could potentially extract the `&Path` view we need, convert that to a `PyObject`, and encode. This feels much simpler than iterating over path parts and encoding each.

Also, I'd prefer if any path encoding code existed in the `conversion` module. Getting that correct is difficult and it is generic code that IMO should live in its own function.

Reply:
## Simplifications

Taking your suggestion of assuming that all possible path separators are `/` and `\` and converting the `path: &Path` to a `PyObject` in one go, I came up with the an implementation (see diff below the fold) that differs mostly in that it no longer takes advantage of `std::path::Iter`'s already knowing what the path separators are. Now, the new code *is* slightly shorter, but it hard codes `/` and `\` and iterates through the resulting `String` twice on Windows instead of once. Maybe using `str.replace` requires less re-allocation of memory than `Vec.push`?

<details><summary>Diff showing new implementation taking your advice</summary>

```diff
diff --git a/pyembed/src/importer.rs b/pyembed/src/importer.rs
index 2325480a..194a14ef 100644
--- a/pyembed/src/importer.rs
+++ b/pyembed/src/importer.rs
@@ -1098,37 +1098,34 @@ impl _PathEntryFinder {
     // Transform b"pkg/mod" into "pkg.mod" on behalf of `OxidizedFinder::path_hook`.
     // Raise an `ImportError` if decoding `path` fails.
     fn parse_path_to_pkg(py: Python, path: &std::path::Path) -> PyResult<String> {
-        let mut pkg = Vec::new();
-        for part in path.iter() {
-            pkg.push(
-                // Use Python's filesystem encoding to ensure correctly
-                // round-tripping the str Python package name from the
-                // potentially non-Unicode path.
-                // https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_170
-                // https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file
-                crate::conversion::path_to_pyobject(py, std::path::Path::new(part))?
-                    .cast_as::<PyString>(py)?
-                    .to_string(py)
-                    .map_err(|mut unicode_err| {
-                        // Need to raise an ImportError if can't decode. For
-                        // clarity, we also attach the UnicodeDecodeError.
-                        // https://docs.python.org/3/reference/import.html#path-entry-finders
-                        let mut imp_err = PyErr::new::<ImportError, _>(py, "cannot decode path");
-                        let imp_exc = imp_err.instance(py);
-                        if let Err(err) = imp_exc.setattr(py, "__suppress_context__", true) {
-                            err
-                        } else if let Err(err) =
-                            imp_exc.setattr(py, "__cause__", unicode_err.instance(py))
-                        {
-                            err
-                        } else {
-                            imp_err
-                        }
-                    })?
-                    .into_owned(),
-            );
-        }
-        Ok(pkg.join("."))
+        // Use Python's filesystem encoding to ensure correctly
+        // round-tripping the str Python package name from the
+        // potentially non-Unicode path.
+        // https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_170
+        // https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file
+        let pkg = crate::conversion::path_to_pyobject(py, path)?
+            .cast_as::<PyString>(py)?
+            .to_string(py)
+            .map_err(|mut unicode_err| {
+                // Need to raise an ImportError if can't decode. For
+                // clarity, we also attach the UnicodeDecodeError.
+                // https://docs.python.org/3/reference/import.html#path-entry-finders
+                let mut imp_err = PyErr::new::<ImportError, _>(py, "cannot decode path");
+                let imp_exc = imp_err.instance(py);
+                if let Err(err) = imp_exc.setattr(py, "__suppress_context__", true) {
+                    err
+                } else if let Err(err) =
+                    imp_exc.setattr(py, "__cause__", unicode_err.instance(py))
+                {
+                    err
+                } else {
+                    imp_err
+                }
+            })?
+            .replace("/", ".");
+        #[cfg(windows)]
+        let pkg = pkg.replace("\\", ".");
+        Ok(pkg)
     }
 }

 ```

</details>

## Extracting something to the `conversion` module

It's not clear to me what part of the function to extract into the `conversion` module. Sixteen of the lines (including comments and braces) of this function in both the old and new implementations do nothing but raise an [`ImportError`] from a `UnicodeDecodeError`. This "conversion" is highly [specific to path hooks](https://docs.python.org/3/reference/import.html#path-entry-finders):
> The hook should expect either a string or bytes object; the encoding of bytes objects is up to the hook (e.g. it may be a file system encoding, UTF-8, or something else), and if the hook cannot decode the argument, it should raise [`ImportError`].

The import documentation makes no other mention of text encodings. To me, this means raising an [`ImportError`] from a `UnicodeDecodeError` need not happen anywhere else in the code base.

In the new implementation, this leaves seven other lines that aren't comments, one of which returns from the function, and three of which replace `/` and `\` with `.`.

The only other thing I can think to change in the `conversion` module is to change the return type of `path_to_pyobject` to be [`PyResult`](http://dgrunwald.github.io/rust-cpython/doc/cpython/type.PyResult.html)`<`[`PyString`](http://dgrunwald.github.io/rust-cpython/doc/cpython/struct.PyString.html)`>` to avoid the `.cast_as::<PyString>(py)?` line in `parse_path_to_pkg`. Had I written `path_to_pyobject` from scratch, maybe* that's the return type I would have picked to begin with? But such a change seems out of scope for this PR. I'm happy to make the change if you disagree.

<details><summary>*</summary>

Not necessarily. I could see an argument for going with `PyBytes` over `PyString`. But again, this strikes me as a debate for a different issue or PR.

</details>

[`ImportError`]: https://docs.python.org/3/library/exceptions.html#ImportError

Reply:
7e2411bbc7640898fd9dbd97dc9fb9ad7156545e

Reply:
Oh, I should mention one more thing here: The `raise ImportError() from unicode_decode_error` code path is currently untested because [the test](https://github.com/indygreg/PyOxidizer/pull/343/files#diff-8be82d8159d7c11fdb6e5fe044a1092032f151c0c8ff57e1bf4dff3babc64324R275-R284) for it fails with a `SystemError` due to dgrunwald/rust-cpython#246. That bug has now been patched but won't be available at least until the next release of rust-cpython.

Reply:
I'll have to think about what we should do with the path hooks conversion. Path encoding is hard. Python gets some of it wrong. I'd prefer to not do that with PyOxidizer. But I have the feeling that compatibility with Python is probably more important here.

Anyway, I like the new proposed code in your comment above. I'll think about it more during the next review. And if it is something that I want to change, I can do that in flight or as a follow-up: I don't want you to worry too much about nuanced issue.

Reply:
Personally, I'm happy with the new code. Frankly, I'm not sure there's much more we can do with it [until we can test it](https://github.com/indygreg/PyOxidizer/pull/343#discussion_r585920650).